File: apps/visualizer/backend/utils/responses.py
"""Standardized response utilities for Flask routes."""
from flask import jsonify
from constants.errors import (
    ERROR_IMAGE_NOT_FOUND,
    ERROR_IMAGE_FILE_NOT_FOUND,
    ERROR_DB_NOT_FOUND,
    ERROR_INTERNAL_SERVER,
)


class _MockResponse:
    """Mock response for testing outside Flask context."""
    def __init__(self, data_dict):
        self._json = data_dict

    @property
    def json(self):
        return self._json


def _make_json_response(data, status_code):
    """Create a JSON response that works with or without Flask app context."""
    try:
        return jsonify(data), status_code
    except RuntimeError:
        # Outside Flask context, return tuple for testing
        return _MockResponse(data), status_code


def error_not_found(resource_type='resource'):
    """Return 404 error for resource not found."""
    messages = {
        'image': ERROR_IMAGE_NOT_FOUND,
        'media': ERROR_IMAGE_NOT_FOUND,
        'database': ERROR_DB_NOT_FOUND,
        'file': ERROR_IMAGE_FILE_NOT_FOUND,
    }
    return _make_json_response(
        {'error': messages.get(resource_type, f'{resource_type} not found')},
        404
    )


def error_bad_request(message='Invalid request'):
    """Return 400 error."""
    return _make_json_response({'error': message}, 400)


def error_server_error(message=None):
    """Return 500 error."""
    return _make_json_response(
        {'error': message or ERROR_INTERNAL_SERVER},
        500
    )


def success_paginated(data, total, offset, limit):
    """Return standardized paginated response."""
    has_more = (offset + limit) < total
    current_page = (offset // limit) + 1
    total_pages = (total + limit - 1) // limit

    return _make_json_response({
        'total': total,
        'data': data,
        'pagination': {
            'offset': offset,
            'limit': limit,
            'current_page': current_page,
            'total_pages': total_pages,
            'has_more': has_more,
        }
    }, 200)

File: apps/visualizer/backend/src/utils/responses.ts
/**
 * Standardized response helpers.
 *
 * Routes must shape JSON through these helpers, not hand-rolled bodies.
 */
import type { Context } from 'hono';
import {
  ERROR_DB_NOT_FOUND,
  ERROR_IMAGE_FILE_NOT_FOUND,
  ERROR_IMAGE_NOT_FOUND,
  ERROR_INTERNAL_SERVER,
} from '../constants/errors.js';

/**
 * An error response whose status the route deliberately does not declare in OpenAPI.
 *
 * Throwing keeps the published contract unchanged while `app.onError` renders it.
 */
export class HttpError extends Error {
  readonly status: 400 | 401 | 403 | 404 | 409 | 429 | 503;

  constructor(status: HttpError['status'], message: string) {
    super(message);
    this.name = 'HttpError';
    this.status = status;
  }
}

export type ResourceType = 'image' | 'media' | 'database' | 'file' | (string & {});

const NOT_FOUND_MESSAGES: Record<string, string> = {
  image: ERROR_IMAGE_NOT_FOUND,
  media: ERROR_IMAGE_NOT_FOUND,
  database: ERROR_DB_NOT_FOUND,
  file: ERROR_IMAGE_FILE_NOT_FOUND,
};

export function errorNotFound(c: Context, resourceType: ResourceType = 'resource') {
  const message = NOT_FOUND_MESSAGES[resourceType] ?? `${resourceType} not found`;
  return c.json({ error: message }, 404);
}

export function errorBadRequest(c: Context, message = 'Invalid request') {
  return c.json({ error: message }, 400);
}

export function errorServerError(c: Context, message?: string) {
  return c.json({ error: message ?? ERROR_INTERNAL_SERVER }, 500);
}

export interface Pagination {
  offset: number;
  limit: number;
  current_page: number;
  total_pages: number;
  has_more: boolean;
}

export interface PaginatedBody<T> {
  total: number;
  data: T[];
  pagination: Pagination;
}

/** Build the paginated envelope. Kept pure so it can be unit-tested without a Context. */
export function paginatedBody<T>(
  data: T[],
  total: number,
  offset: number,
  limit: number,
): PaginatedBody<T> {
  return {
    total,
    data,
    pagination: {
      offset,
      limit,
      current_page: Math.floor(offset / limit) + 1,
      total_pages: Math.floor((total + limit - 1) / limit),
      has_more: offset + limit < total,
    },
  };
}

export function successPaginated<T>(
  c: Context,
  data: T[],
  total: number,
  offset: number,
  limit: number,
) {
  return c.json(paginatedBody(data, total, offset, limit), 200);
}

const DEFAULT_LIMIT = 50;
const MAX_LIMIT = 500;

/**
 * Clamp `limit`/`offset`: unparseable values fall back to defaults rather than erroring.
 */
export function clampPagination(
  limit: unknown,
  offset: unknown,
  defaultLimit = DEFAULT_LIMIT,
): { limit: number; offset: number } {
  const toInt = (v: unknown, fallback: number): number => {
    if (v === null || v === undefined || v === '') return fallback;
    const n = Number(v);
    return Number.isFinite(n) ? Math.trunc(n) : fallback;
  };
  return {
    limit: Math.max(1, Math.min(MAX_LIMIT, toInt(limit, defaultLimit))),
    offset: Math.max(0, toInt(offset, 0)),
  };
}
